pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state/products/apmsampling"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/log"
+	"github.com/DataDog/datadog-agent/pkg/trace/sampler"
+)
+
+type apmsamplingRemoteConfigID string
+
+const (
+	dynamicRatesByService     apmsamplingRemoteConfigID = "dynamic_rates"
+	userDefinedRatesByService apmsamplingRemoteConfigID = "user_rates"
+	samplerconfig             apmsamplingRemoteConfigID = "samplerconfig"
+)
+
+type prioritySampler interface {
+	UpdateTargetTPS(targetTPS float64)
+	UpdateRemoteRates(updates []sampler.RemoteRateUpdate)
+}
+
+type errorsSampler interface {
+	UpdateTargetTPS(targetTPS float64)
+}
+
+type rareSampler interface {
+	SetEnabled(enabled bool)
+}
+
+// RemoteConfigHandler holds pointers to samplers that need to be updated when APM remote config changes
+type RemoteConfigHandler struct {
+	remoteClient          config.RemoteClient
+	prioritySampler       prioritySampler
+	errorsSampler         errorsSampler
+	rareSampler           rareSampler
+	agentConfig           *config.AgentConfig
+	remoteConfigPathRegex *regexp.Regexp
+}
+
+func New(conf *config.AgentConfig, prioritySampler prioritySampler, rareSampler rareSampler, errorsSampler errorsSampler) *RemoteConfigHandler {
+	if conf.RemoteSamplingClient == nil {
+		return nil
+	}
+
+	// paths are in the form datadog/<org ID>/<product name>/<config ID>/config
+	r, err := regexp.Compile(`datadog\/\d+\/APM_SAMPLING\/(.+)\/config`)
+	if err != nil {
+		panic(err)
+	}
+
+	return &RemoteConfigHandler{
+		remoteClient:          conf.RemoteSamplingClient,
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		if configID == dynamicRatesByService || configID == userDefinedRatesByService {
```

Comment:
nit: This would probably read easier as a `switch` and make it easier to add more `configID` down the line

Reply:
I agree, fixed